### PR TITLE
External Linking Support

### PR DESCRIPTION
**Pull request by solocommand**
Backend and Frontend support for setting external links on orders and campaigns.

screen shot 2019-01-23 at 9 41 23 am
screen shot 2019-01-23 at 9 41 01 am
screen shot 2019-01-23 at 9 40 43 am
screen shot 2019-01-23 at 9 38 59 am
screen shot 2019-01-23 at 9 42 53 am

### CODE REVIEW

**Comment by zarathustra323** [comment]
@solocommand What is the purpose of the blue buttons? They're just mimicking the link title?

Other design things:
1. The add link should be green and be the same size and use a plus icon and a link icon
2. Can we just remove the blue button all together?
3. Can the title field, url field, and remove buttons be combined into one line as an input group?

File: services/graphql/src/graphql/definitions/lineitem.js [hunk 1]
```diff
@@ -50,6 +51,7 @@ type LineItem implements Timestampable & UserAttributable @applyInterfaceFields
   dates: LineItemDates!
   priority: Int!
   ads(input: LineItemAdsInput = {}): AdConnection! @refMany(modelName: "ad", localField: "_id", foreignField: "lineitemId")
+  externalLinks: [ExternalLink]
```
Comment (zarathustra323) [inline services/graphql/src/graphql/definitions/lineitem.js:54]: Can this be changed to `[ExternalLink!]!` and then ensure that the `externalLinks` field always resolves to an array?

File: services/graphql/src/graphql/definitions/order.js [hunk 2]
```diff
@@ -18,6 +18,7 @@ extend type Mutation {
 
   orderName(input: OrderNameMutationInput!): Order! @requiresAuth @setAndUpdate(modelName: "order", path: "name")
   orderAdvertiser(input: OrderAdvertiserMutationInput!): Order! @requiresAuth @setAndUpdate(modelName: "order", path: "advertiserId")
+  orderExternalLinks(input: ExternalLinksMutationInput): Order! @requiresAuth @setAndUpdate(modelName: "order", path: "externalLinks")
```
Comment (zarathustra323) [inline services/graphql/src/graphql/definitions/order.js:21]: Is there a reason to have links on both the order and the line item?

File: services/graphql/src/mongoose/plugins/external-linking.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  label: {
+    type: String,
+    required: false,
+    trim: true,
+  },
+});
+
+module.exports = function externalLinkingPlugin(schema) {
+  schema.add({
+    externalLinks: [LinkSchema],
```
Comment (zarathustra323) [inline services/graphql/src/mongoose/plugins/external-linking.js:29]: Add `default: []`

**Comment by solocommand** [comment]
> @zarathustra323 What is the purpose of the blue buttons? They're just mimicking the link title?

It's the actual link you can click on

> 1. The add link should be green and be the same size and use a plus icon and a link icon

ok

> 2. Can we just remove the blue button all together?

Need to click

> 3. Can the title field, url field, and remove buttons be combined into one line as an input group?

That's how i had it originally but it didn't work well at multiple resolutions. works better as individual cols (the input group doesn't resize)

File: services/manage/app/components/external-links.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    if (this.get('shouldSave')) {
+      debounce(this, 'sendOnChange', value, event, 1, true);
+    } else {
+      this.sendOnChange(value, event);
+    }
+  },
+
+  actions: {
+    createLink() {
+      this.get('value').pushObject({ label: 'New Link', url: '' });
```
Comment (zarathustra323) [inline services/manage/app/components/external-links.js:66]: The `label` should be an empty string by default... it's more difficult for someone to have to delete `New Link` every time as opposed to just leaving it blank

File: services/manage/app/components/external-links.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  /**
+   * Debounces the input onchange event.
+   *
+   * Will immediately fire the change, if dirty, and
+   * will cancel any other pending change events.
+   *
+   * @param {Event} event
+   */
+  debounceChange(event) {
```
Comment (zarathustra323) [inline services/manage/app/components/external-links.js:54]: this doesn't need to be debounced

File: services/manage/app/components/external-links.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
+import Component from '@ember/component';
+import { isArray } from '@ember/array';
+import { get, set } from '@ember/object';
+import { debounce } from '@ember/runloop';
+
+export default Component.extend({
+  isModalShowing: false,
+
+  isCreating: false,
```
Comment (zarathustra323) [inline services/manage/app/components/external-links.js:9]: I don't think `isModalShowing` `isCreating` or `typeDelay` are needed

File: services/manage/app/templates/manage/orders/edit.hbs [hunk 7]
```diff
@@ -26,6 +26,11 @@
           {{lineitem/-icon}} Line Items <span class="badge badge-light">{{model.lineitems.totalCount}}</span>
         {{/link-to}}
       </li>
+      <li class="nav-item">
+        {{#link-to "manage.orders.edit.links"  class="nav-link" role="tab"}}
+          {{entypo-icon "link"}} Links
```
Comment (zarathustra323) [inline services/manage/app/templates/manage/orders/edit.hbs:31]: Turn this into an `external-link/-icon` component to be consistent with the other icon comps

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

**Comment by zarathustra323** [comment]
@solocommand You can make the URL clickable without a huge, blue button :)

My point is (and sorry for any offense) the layout is very blocky and uses up a lot of real estate. I just think there's a way to shore it up and present it better

File: services/manage/app/templates/components/external-links.hbs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    </div>
+  {{/auto-save/form}}
+{{/each}}
+
+<div class="mt-1 text-right">
+  <button class="btn btn-sm btn-primary" type="button" {{action "createLink"}}>{{entypo-icon "plus"}} Add Link</button>
+</div>
+
+{{#if error}}
+  <div class="my-1 alert alert-danger" role="alert">
```
Comment (zarathustra323) [inline services/manage/app/templates/components/external-links.hbs:69]: Errors should be handled by each input like other form elements

**Review by zarathustra323 — CHANGES_REQUESTED** [review]

File: services/manage/app/templates/components/external-links.hbs (hunk 8 above)
Comment (solocommand) [reply services/manage/app/templates/components/external-links.hbs:69]: The individual inputs aren't validated/saved though. The whole array of links is sent at once

Comment (zarathustra323) [reply services/manage/app/templates/components/external-links.hbs:69]: Should it?

File: services/graphql/src/graphql/definitions/order.js (hunk 2 above)
Comment (solocommand) [reply services/graphql/src/graphql/definitions/order.js:21]: It was requested that way

**Pull request closed without merging**